### ISSUE TITLE
[java] PrematureDeclaration as result of method call (false positive)

### ISSUE THREAD

**Issue by ben-manes** [pmd/pmd#1429]
**Affects PMD Version:** 
6.9.0

**Rule:**
[PrematureDeclaration](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#prematuredeclaration)

**Description:**
The result of a `ConcurrentMap#put` is captured for handling later in the method. As `put` has a side effect, the logic changes if the declaration is moved downward.

**Code Sample demonstrating the issue:**
```java
var prior = cache.put(uri, cachedPath);

if (ref == null) {
  return PathRef.create(cachedPath, 1);
}
if (prior == null) {
  ref.count().incrementAndGet();
}
```

**Comment by jsotuyod** [comment]
@ben-manes thanks for the report. Unfortunately, as you said, this requires PMD knowing that `Map.put` has side effects.

To do so, PMD would either have to look into `Map` itself (we don't have the source, only the binary classes, and again, `Map` is an interface); or we would have to somehow whitelist this usage (which may be hard to maintain as the list of methods with side effects may be indefinitely long).

For the time being I don't think we have a better fix than to suppress this warning, but I'll keep it open for future reference. Any ideas on better ways to work this out are welcomed.

**Comment by Andy-2639** [comment]
I think this is related to #1208

Maybe all method calls with unknown behavior (as far as I understand this means literally all method calls) should be excluded from the premature declaration rule.

**Comment by Fernal73** [comment]
> @ben-manes thanks for the report. Unfortunately, as you said, this requires PMD knowing that `Map.put` has side effects.
> 
> To do so, PMD would either have to look into `Map` itself (we don't have the source, only the binary classes, and again, `Map` is an interface); or we would have to somehow whitelist this usage (which may be hard to maintain as the list of methods with side effects may be indefinitely long).
> 
> For the time being I don't think we have a better fix than to suppress this warning, but I'll keep it open for future reference. Any ideas on better ways to work this out are welcomed.

The only way to do it without going into byte-code analysis to write a custom annotation (say SideEffectFree) and annotate the method call with it. PMD just has to check this annotation. Should this be a custom annotation with Simple Name SideEffectFree or should PMD provide its own set of annotations to be used by developers that work well with its rules engine?

**Comment by Fernal73** [comment]
Another code sample where PrematureDeclaration is displaying False-positive with methods that have side-effects...

```java
package ds; 

import java.io.BufferedInputStream; 
import java.io.File;
 import java.io.IOException; 
import java.io.InputStream;
 import java.net.Socket; 
import java.net.URL; 
import java.net.URLConnection; 
import java.nio.file.Files; 

import java.util.NoSuchElementException;

 public final class BinaryInputStream { 
private static final int EOF = -1; // end of file private static final String READING_EMPTY = "Reading from empty input stream"; private BufferedInputStream in; 
// the input stream 
private int buffer;
 // one character buffer 
private int n; 
// number of bits left in buffer
 

 public BinaryInputStream() {
 in = new BufferedInputStream(System.in); fillBuffer(); 
} 

 private void fillBuffer() { 
try { buffer = in.read(); n = 8; } catch (IOException e) { System.err.println("EOF : " + e.getMessage()); buffer = EOF; n = -1; } } 

 public boolean isEmpty() { return buffer == EOF; } 

public char readChar() { 
if (isEmpty()) throw new NoSuchElementException(READING_EMPTY); // special case when aligned byte 
if (n == 8) {
 int x = buffer; 
fillBuffer();
 return (char) (x & 0xff); 
} 
// combine last N bits of current buffer with first 8-N bits of new buffer 
int x = buffer; x <<= 8 - n; 
int oldN = n; // flagged for PrematureDeclaration

fillBuffer();
 if (isEmpty()) throw new NoSuchElementException(READING_EMPTY); n = oldN; 
x |= buffer >>> n;
 return (char) (x & 0xff); 
// the above code doesn't quite work for the last character if N = 8 
// because buffer will be -1 
} 

  } 
```

**Comment by adangel** [comment]
This has been fixed with PMD 7.0.0-rc1.
